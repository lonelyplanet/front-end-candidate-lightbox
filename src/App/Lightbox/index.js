import React, { Component } from 'react'
import 'whatwg-fetch'
import './index.scss'

const API_KEY = 'MTgwNzAw'

class Lightbox extends Component {
  constructor (props) {
    super(props)

    this.state = { url: '', loading: false, error: '' }

    this.handleKeydown = this.handleKeydown.bind(this)
    this.handleClick = this.handleClick.bind(this)
    this.handleCloseBtn = this.handleCloseBtn.bind(this)
  }

  componentWillMount () {
    this.setState({ loading: true })

    fetch(`http://thecatapi.com/api/images/get?api_key=${API_KEY}`)
      .then((response) => {
        this.setState({ url: response.url, loading: false })
      })
      .catch(() => {
        this.setState({ loading: false, error: 'Unable to get a cat. Try again' })
      })

    document.addEventListener('keydown', this.handleKeydown);
    document.addEventListener('click', this.handleClick);
  }

  componentWillUnmount () {
    document.removeEventListener('keydown', this.handleKeydown);
    document.removeEventListener('click', this.handleClick);
  }

  handleClick (event) {
    if (event.target.className == 'lightbox') {
      this.props.onClose()
    }
  }

  handleCloseBtn (event) {
    event.preventDefault()
    this.props.onClose()
  }

  handleKeydown (event) {
    if (event.keyCode === 27) {
      this.props.onClose()
    }
  }

  render () {
    const loadingMessage = this.state.loading
      ? (<h3>Loading Catâ€¦</h3>)
      : null

    const errorMessage = this.state.error
      ? (<h4>{this.state.error}</h4>)
      : null

    return (
      <div className='lightbox'>
        <div className='lightbox-content'>
          <img src={this.state.url} />
          {loadingMessage}
          {errorMessage}
          <a href className="button" onClick={this.handleCloseBtn}>Close</a>
        </div>
      </div>
    )
  }
}

export default Lightbox
