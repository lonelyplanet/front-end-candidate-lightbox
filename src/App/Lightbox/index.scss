.lightbox {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background: rgba(0, 0, 0, 0.2);
  display: flex;
  justify-content: center;
  align-items: center;
}

.lightbox-content {
  background: #ffffff;
  width: 30vw;
  padding: 1rem;

  img {
    max-width: 100%;
    display: block;
    margin: 0 auto 1rem;
  }
}