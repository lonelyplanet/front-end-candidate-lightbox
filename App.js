import React, { Component } from 'react';
import Lightbox from 'react-image-lightbox';

const images = [
       'https://www.jpl.nasa.gov/edu/images/news/cassini_ringcrossing.gif',
       'http://www.slate.com/content/dam/slate/blogs/bad_astronomy/2014/06/Ten%20Years%20at%20Saturn/cassini_backlit.jpg.CROP.original-original.jpg',
       'http://static4.businessinsider.com/image/58e5587677bb70f3398b6506-800',
       'https://www.nasa.gov/sites/default/files/thumbnails/image/cassini_grandfinale_orbits.jpg',
       'http://www.aljazeera.com/mritems/Images/2017/4/26/e918aee1b243426d8fcccb1e65c3e902_18.jpg',
       'https://saturn.jpl.nasa.gov/internal_resources/791/',
       'http://a.abcnews.com/images/Technology/HT-cassini-nasa-5804_PIA14944_1x1_1600.jpg',
       'https://3c1703fe8d.site.internapcdn.net/newman/gfx/news/hires/2017/2-cassinimissi.jpg',
  ];


class App extends React.Component {
    constructor(props) {
        super(props);
 
        this.state = {
            photoIndex: 0,
            isOpen: false
        };
    }
 
    render() {
        const {
            photoIndex,
            isOpen,
        } = this.state;
 
        return (
            <div>
                <button
                    type="button"
                    className="button js-lightbox-opener"
                    onClick={() => this.setState({ isOpen: true })}
                >
                    Open Lightbox to see Cassini images
                </button>
 
                {isOpen &&
                    <Lightbox
                        mainSrc={images[photoIndex]}
                        nextSrc={images[(photoIndex + 1) % images.length]}
                        prevSrc={images[(photoIndex + images.length - 1) % images.length]}
 
                        onCloseRequest={() => this.setState({ isOpen: false })}
                        onMovePrevRequest={() => this.setState({
                            photoIndex: (photoIndex + images.length - 1) % images.length,
                        })}
                        onMoveNextRequest={() => this.setState({
                            photoIndex: (photoIndex + 1) % images.length,
                        })}
                        
                    />
                }
            </div>
        );
    }
}
export default App;