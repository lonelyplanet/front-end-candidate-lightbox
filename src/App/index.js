import React, { Component } from 'react'
import Lightbox from './Lightbox'
import './index.scss'

class App extends Component {
  constructor (props) {
    super(props)

    this.state = { lightboxOpen: false }

    this.openLightbox = this.openLightbox.bind(this)
    this.closeLightbox = this.closeLightbox.bind(this)
  }

  openLightbox () {
    this.setState({ lightboxOpen: true })
  }

  closeLightbox () {
    this.setState({ lightboxOpen: false })
  }

  renderLightbox () {
    if (this.state.lightboxOpen) {
      return <Lightbox onClose={this.closeLightbox} />
    }
  }

  render () {
    return (
      <div>
        <a href='#' className='button' onClick={this.openLightbox}>open lightbox</a>
        {this.renderLightbox()}
      </div>
    )
  }
}

export default App
