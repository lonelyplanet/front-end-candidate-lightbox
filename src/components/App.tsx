import * as React from 'react'
import Lightbox from '~/components/Lightbox'

interface AppProps {}
interface AppState {
  showLightbox: boolean
}
class App extends React.Component<AppProps, AppState> {
  state: AppState = {
    showLightbox: false,
  }

  _onShowLightbox = () => {
    this.setState({ showLightbox: true })
  }

  _onCloseLightbox = () => {
    this.setState({ showLightbox: false })
  }

  render () {
    return (
      <div>
        <h1>Hello Lonely Planet</h1>
        <button onClick={this._onShowLightbox}>Open Demo Lightbox</button>
        {this.state.showLightbox && (
          <Lightbox onClose={this._onCloseLightbox} />
        )}
      </div>
    )
  }
}

export default App
