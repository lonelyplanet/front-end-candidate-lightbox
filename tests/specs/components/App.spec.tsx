import * as React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import App from '~/components/App'

test('matches snapshot', () => {
  expect(toJson(shallow(<App />))).toMatchSnapshot()
})
