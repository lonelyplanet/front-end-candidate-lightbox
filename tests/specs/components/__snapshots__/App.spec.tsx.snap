// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<h1>
  Hello Lonely Planet
</h1>
`;
